File: patterns/07_command/python/mandatum.py
#!/usr/bin/env python3
"""
COMMAND PATTERN — Imperial Building Commands
"Mandatum datum, mandatum executum"

Run: python3 mandatum.py
"""
from abc import ABC, abstractmethod
from typing import List

class RomeArchitect:
    def __init__(self): self.buildings: List[str] = []
    def build(self, s: str) -> str:
        self.buildings.append(s)
        return f"🏛  {s} BUILT! {self.buildings}"
    def demolish(self, s: str) -> str:
        if s in self.buildings: self.buildings.remove(s)
        return f"🔨 {s} DEMOLISHED! {self.buildings}"

class ImperialCommand(ABC):
    @abstractmethod
    def execute(self) -> str: ...
    @abstractmethod
    def undo(self)    -> str: ...

class BuildCmd(ImperialCommand):
    def __init__(self, arch: RomeArchitect, building: str):
        self._arch, self._building = arch, building
    def execute(self) -> str: return self._arch.build(self._building)
    def undo(self)    -> str: return self._arch.demolish(self._building)

class ImperialPalace:
    def __init__(self):
        self._history: List[ImperialCommand] = []
        self._undone:  List[ImperialCommand] = []
    def issue(self, cmd: ImperialCommand) -> str:
        r = cmd.execute(); self._history.append(cmd); self._undone.clear(); return r
    def undo_last(self) -> str:
        if not self._history: return "Nothing to undo!"
        cmd = self._history.pop(); self._undone.append(cmd); return cmd.undo()
    def redo_last(self) -> str:
        if not self._undone: return "Nothing to redo!"
        cmd = self._undone.pop(); self._history.append(cmd); return cmd.execute()

def main():
    print("╔═══════════════════════════════════════════════╗")
    print("║   COMMAND PATTERN — Python                    ║")
    print("║   Imperial Building Commands                  ║")
    print("╚═══════════════════════════════════════════════╝\n")
    arch   = RomeArchitect()
    palace = ImperialPalace()
    print("── ISSUING COMMANDS ────────────────────────────")
    print(palace.issue(BuildCmd(arch, "Colosseum")))
    print(palace.issue(BuildCmd(arch, "Aqua Claudia")))
    print(palace.issue(BuildCmd(arch, "Temple of Jupiter")))
    print("\n── VESPASIAN CHANGES HIS MIND ──────────────────")
    print(palace.undo_last())
    print(palace.undo_last())
    print("\n── AND BACK AGAIN ──────────────────────────────")
    print(palace.redo_last())
    print('\n"Mandatum datum, mandatum executum!"')

if __name__ == "__main__":
    main()

File: patterns/07_command/java/Mandatum.java
// ============================================================
//  COMMAND PATTERN — Imperial Building Commands
//  Run: java Mandatum.java
// ============================================================
import java.util.*;

public class Mandatum {

    static class RomeArchitect {
        final List<String> buildings = new ArrayList<>();
        String build(String s) {
            buildings.add(s);
            return "🏛  " + s + " BUILT! " + buildings;
        }
        String demolish(String s) {
            buildings.remove(s);
            return "🔨 " + s + " DEMOLISHED! " + buildings;
        }
    }

    interface ImperialCommand { String execute(); String undo(); }

    static class BuildCmd implements ImperialCommand {
        private final RomeArchitect arch;
        private final String building;
        BuildCmd(RomeArchitect a, String b) { this.arch = a; this.building = b; }
        public String execute() { return arch.build(building); }
        public String undo()    { return arch.demolish(building); }
    }

    static class ImperialPalace {
        private final Deque<ImperialCommand> history = new ArrayDeque<>();
        private final Deque<ImperialCommand> undone  = new ArrayDeque<>();

        String issue(ImperialCommand cmd) {
            String r = cmd.execute();
            history.push(cmd); undone.clear();
            return r;
        }
        String undoLast() {
            if (history.isEmpty()) return "Nothing to undo!";
            var cmd = history.pop(); undone.push(cmd);
            return cmd.undo();
        }
        String redoLast() {
            if (undone.isEmpty()) return "Nothing to redo!";
            var cmd = undone.pop(); history.push(cmd);
            return cmd.execute();
        }
    }

    public static void main(String[] args) {
        System.out.println("╔═══════════════════════════════════════════════╗");
        System.out.println("║   COMMAND PATTERN — Java                      ║");
        System.out.println("║   Imperial Building Commands                  ║");
        System.out.println("╚═══════════════════════════════════════════════╝\n");

        var arch   = new RomeArchitect();
        var palace = new ImperialPalace();

        System.out.println("── ISSUING COMMANDS ────────────────────────────");
        System.out.println(palace.issue(new BuildCmd(arch, "Colosseum")));
        System.out.println(palace.issue(new BuildCmd(arch, "Aqua Claudia")));
        System.out.println(palace.issue(new BuildCmd(arch, "Temple of Jupiter")));

        System.out.println("\n── UNDO ─────────────────────────────────────────");
        System.out.println(palace.undoLast());
        System.out.println(palace.undoLast());

        System.out.println("\n── REDO ─────────────────────────────────────────");
        System.out.println(palace.redoLast());

        System.out.println("\n\"Mandatum datum, mandatum executum!\"");
    }
}
